src/adc.rs

```diff
@@ -0,0 +1,79 @@
+//! Analog-digital conversion traits
+
+use nb;
+
+/// A marker trait to identify MCU pins that can be used as inputs to an ADC channel.
+///
+/// This marker trait denotes an object, i.e. a GPIO pin, that is ready for use as an input to the
+/// ADC. As ADCs channels can be supplied by multiple pins, this trait defines the relationship
+/// between the physical interface and the ADC sampling buffer.
+pub trait AdcChannel {
```

Comment:
I think we should rename this to `Channel`. The module is already called `adc`, and I think `adc::Channel` is much nicer than `adc::AdcChannel`.

Reply:
In my opinion trait better name `AdcChannel` and add it to prelude.
`Channel` name better reserve for `enum` of existent channels.

It would be something like this:
For 103xx:
```

pub trait AdcChannel {
    /// Channel ID type
    ///
    /// A type used to identify this ADC channel. For example, if the ADC has eight channels, this
    /// might be a `u8`. If the ADC has multiple banks of channels, it could be a tuple, like
    /// `(u8: bank_id, u8: channel_id)`.
    type ID;
    /// The specific ID that identifies this channel, for example `0` for the first ADC channel.
    const CHANNEL: Self::ID;

    fn channel(&self) -> Self::ID {
        Self::CHANNEL
    }
}

/// ADC_channels
#[derive(Clone, Copy, PartialEq)]
pub enum Channel {
    /// ADC Channel 0 selected
    C0 = 0x00,
    /// ADC Channel 1 selected
    C1 = 0x01,
    ...
    /// ADC Channel 15 selected
    C15 = 0x0F,
    /// ADC Channel 16 selected
    Temperature = 0x10,
    /// ADC Channel 17 selected
    Voltage = 0x11,
}

macro_rules! analog_input {
    ($PXi:ident, $CHx:ident, $SMPRx:ident) => {
        impl AdcChannel for $PXi<Analog> {
            type ID = Channel;
            const CHANNEL: Channel = Channel::$CHx;
        }
    };
}
analog_input!(PA0, C0);
```

Reply:
@burrbull Traits in the prelude are re-exported under a different name to prevent name collisions, so that shouldn't be an issue. And I image non-prelude use would not be an issue either, as an implementation could do it like this:
``` rust
use embedded_hal::adc;

impl adc::Channel for MyChannel {
    // ...
}

pub enum Channel {
    // ...
}
```